mmgen/models/architectures/sngan_proj/generator_discriminator.py

```diff
@@ -129,33 +121,55 @@ def forward(self, noise, num_batches=0, y=None, return_noise=False):
             assert num_batches > 0
             noise_batch = torch.randn((num_batches, self.noise_size))
 
+        if isinstance(label, torch.Tensor):
+            assert label.ndim == 1, ('The label shoube be in shape of (n, )'
+                                     f'but got {label.shape}.')
+            label_batch = label
+        elif callable(label):
+            label_generator = label
+            assert num_batches > 0
+            label_batch = label_generator((num_batches))
+        else:
+            assert num_batches > 0
+            label_batch = torch.randint(0, self.num_classes, (num_batches, ))
+
         # dirty code for putting data on the right device
         noise_batch = noise_batch.to(get_module_device(self))
 
         x = self.noise2feat(noise_batch)
         x = x.reshape(x.size(0), -1, self.input_scale, self.input_scale)
 
-        for res_block in self.res_blocks:
-            x = res_block(x, y)
+        for conv_block in self.conv_blocks:
+            x = conv_block(x, label_batch)
 
         out_img = self.to_rgb(x)
 
+        out_dict = dict()
         if return_noise:
-            output = dict(fake_img=out_img, noise_batch=noise_batch)
-            return output
-
-        return out_img
-
-    def init_weight(self, pretrained=False):
-        if pretrained:  # TODO:
-            pass
+            out_dict['noise'] = noise_batch
+        if return_label:
```

Comment:
We may remove `return_label` flag. I suggest that `return_noise` indicates to return the out_dict containing all kinds of datas. When writing the interface, we should be consistent with other modules like `DCGAN generator/discriminator`. Pay attention to the keys in the output dict.

```python

if return_noise:
    out_dict = dict(noise_batch=noise_batch, label=label)
    return out_dict
else:
    return out_image

```

Reply:
`key` for noise term in `out_dict` has fixed. But I still suggest keeping `return_label`, for label and noise are not the same things intuitively.
What's your suggestion here? @plyfager

Reply:
In this initial design, `return_noise` flag indicates to return all of related features or data in `generator` or `discriminator`. Because other modules (like GenerativeVisualizationHook, sample_from_noise) follow this convention to parse the output of generators and discriminators. I may suggest that we should follow this rule. :)

Reply:
> `key` for noise term in `out_dict` has fixed. But I still suggest keeping `return_label`, for label and noise are not the same things intuitively.
> What's your suggestion here? @plyfager

I think users usually want to get the outputs only or all related data, so `return_noise` is enough. However, we may change the flag name for a better one?

Reply:
We can first keep it and change it later.